Support inherited field in object type expression

objmagic opened the pull request:
This PR allows one to inherit closed object type expression when defining object type.

Here is an example. Entering the following code in interpreter
```OCaml
type t = <m:int>
type g = <n:string; t>
type h = <x:string; y:int; g>
``` 
and you will have the following output
```OCaml
type t = < m : int >
type g = < m : int; n : string >
type h = < m : int; n : string; x : string; y : int >
```

This is [an OCaml Labs compiler hacking task](https://github.com/ocamllabs/compiler-hacking/wiki/Things-to-work-on#support-extending-existing-object-types). Also, there was [a thread on caml-list](https://sympa.inria.fr/sympa/arc/caml-list/2017-02/msg00067.html) last month discussing about it.

One small refactor along with this PR is that I added  location on object field label and polymorphic variant tags. See [parsetree.mli#L156](https://github.com/ocaml/ocaml/blob/trunk/parsing/parsetree.mli#L156).

objmagic committed:
Support inherited field in object type expression.

Also add location on object field label and polymorphic variant tags.

objmagic committed:
Merge branch 'trunk' into object-inherit

objmagic committed:
Add another test.

objmagic committed:
Remove message.

objmagic committed:
Add Change entry.

objmagic commented:
<del>One problem I am aware of is that the following code</del>
```OCaml
class type t1 = object method f : int end
type t = < #t1 >
```
<del>will give you error message: "Error: The type #t1 is not an object type".

<del>Not sure if it makes sense to support this pattern though.

Update (04/06):
```OCaml
class type t1 = object method f : int end
type t = < t1 >
```
is supported now

bobzhang commented:
this looks really useful, structural typing are very useful for projects like JSOO and BuckleScript. some bikeshedding about syntax, `g` looks punning, maybe we can it more explicit:

```ocaml
type t = < x : int ; y : int; inherit g >
```

lpw25 commented:
> some bikeshedding about syntax, g looks punning, maybe we can it more explicit

There's no punning anywhere in the type algebra so I doubt it will confuse people. Whereas using `inherit` will likely perpetuate people's misunderstandings about the relation between class and object types. Inheritance is about inheriting behaviour, whilst it can cause the object type associated with a class to be extended it is not the same thing and I don't think it should use the same syntax.

alainfrisch commented:
> Inheritance is about inheriting behaviour

`inherit` is also used in class types, where it doesn't imply any inheritance of behavior:

````ocaml
class type c = object method x: int end
class type d = object inherit c method y: int end
````

bluddy commented:
So maybe it's more accurate to say that `inherit` is associated with nominal typing rather than structural, and we don't want to confuse the two?

lpw25 commented:
> inherit is also used in class types, where it doesn't imply any inheritance of behavior

I know and it's kind of wrong there too. The class system has quite a few things like this which subtly confuse different concepts, and I've always thought that it was one of the reasons people have a lot of trouble understanding it. I'd rather not add one more to the list, even if it would not be the worst offender.

yallop commented:
I like the syntax that @objmagic has implemented because it parallels the syntax for polymorphic variants:

```ocaml
  type d = [`A of int | `B of int | c ]
  type d = < a :  int ;  b :  int ; c >
```

Rather less importantly, it's also reminiscent of the functional update syntax for records:

```ocaml
  { r with x = a ; y = b }
  < r ;    x : a ; y : b >
```

gasche commented:
> it parallels the syntax for polymorphic variants

This is the killer argument, and I think it concludes the syntactic discussion.

Any takers for an actual code review? :-)

yallop reviewed (commented):
It'd be good to have a few more tests, e.g. where an extended type is more than just a simple name

```ocaml
   type 'a t = < m: 'a >
   type s = < int t >

   module M = struct type t = < m: int > end
   type u = < M.t >

   type r = < a : int; < b : int > >
```

and where an extended type is empty

```ocaml
   type e = < >
   type r1 = < a : int; e >

   type r2 = < a : int; < > >
```

objmagic committed:
Fix bug when handling empty object type exp

Also add tests.

objmagic committed:
Update Changes entry.

objmagic commented:
@yallop Thanks! These are good tests. I fixed one bug when handling empty object type expression (https://github.com/ocaml/ocaml/pull/1118/commits/7743e96e81341c8410896c3b29236ea511233ef9)

objmagic commented:
@yallop would you like to also have a look at my question at https://github.com/ocaml/ocaml/pull/1118#issuecomment-288623135?

@Octachron would you like to help review [change on odoc](https://github.com/ocaml/ocaml/pull/1118/files#diff-66acf6c92081b9ff3459634a9e186510)? I was thinking to add some tests for that but I do not know how to do that.

garrigue commented:
I've been thinking of adding this features a number of times.
I always stopped short of doing it, because one can already do that using class types.
So this is a "political decision".
I agree that re-establishing the symmetry with polymorphic variants would be nice, but we need to discuss that on `caml-devel`.

For the implementation, I'm a bit concerned by the way `transl_fields` is implemented.
This is clearly not symmetrical with polymorphic variants: rather than unification, fields are copied, and as a result repeated declaration of a field is not allowed. This may be a problem with multiple inheritance. I think that keeping things in line with polymorphic variants would be better; actually the restriction on duplicate field declarations doesn't seem relevant for object types.

objmagic commented:
@garrigue I agree that implementation should be symmetrical with polymorphic variants' implementation. However I don't understand your statement "the restriction on duplicate field declarations doesn't seem relevant for object types". Do you mean `type t = <a: int; a: string>` (not allowed now) should give you `type t = <a:string>`, following the rule "last wins" in class?

garrigue commented:
@objmagic No, this should give you a unification error, like for polymorphic variants. We're talking about objects, not modules.

objmagic committed:
Unify, rather than copy.

objmagic commented:
@garrigue Thank you very much for the comments! I have updated my PR to use unification instead (https://github.com/ocaml/ocaml/pull/1118/commits/d459674326358b19dd2f115a1aa143c3429b87d2). If I understand your words correctly, I removed "restriction on duplicate field declarations" (as a side-effect of using unification)

Octachron commented:
@objmagic On the ocamldoc side, if you want to add tests, there are two kind of tests available in the testsuite. First, basic type printing tests are located in `testsuite/tests/tool-ocamldoc`, in particular `testsuite/tests/tool-ocamldoc/t01.ml` already tests the printing of object types.
Then, the various backend are tested in
 * tool-ocamldoc-2: latex
 * tool-ocamldoc-html: html
 * tool-ocamldoc-man: man pages

Adding an `Object_types.mli` test file in at least one of the back-end tests would probably be a good idea.

Looking at the code and ocamldoc output, currently the information about inherited object type is not used by ocamldoc. Consequently
```OCaml
type t = <a:int (** comment for a *) ; b:int (** comment for b*) >
type r = < t (**comment for inherited t*) ; c:int (** comment for c*) >
``` 
is printed as
```OCaml
type t = <a:int (** comment for a *) ; b:int (** comment for b*) >
type r = < a:int; b:int ; c:int (** comment for c*) >
```
Ideally, I think it would be better to not expand inherited fields in the documentation and add the possibility to comment on inherited fields. However, these are quality of life improvements, and could be the object of a subsequent PR, since ocamldoc is neither crashing nor missing object fields in presence of inherited fields.

Octachron commented on ocamldoc/odoc_sig.ml:
```diff
@@ -282,23 +282,26 @@ module Analyser =
+              | Otag ({txt=name}, _, ct) ::
+                  ((Oinherit ct2) as ele2) :: q
+              | Otag ({txt=name}, _, ct) ::
+                  ((Otag (_, _, ct2)) as ele2) :: q ->
```
It might be worthwhile to fusion the two cases here into 
```OCaml
| (Otag ({txt=name}, _, ct) ) :: ((Oinherit ct2 | Otag (_, _, ct2) ) as ele2) :: q
```

Octachron commented on ocamldoc/odoc_sig.ml:
```diff
@@ -282,23 +282,26 @@ module Analyser =
                 let pos2 = Loc.ptyp_start ct2 in
                 let (_,comment_opt) = just_after_special pos pos2 in
                 (name, comment_opt) :: (f (ele2 :: q))
+              | _ :: q -> f q
```
Note that the `OInherit _` case is essentially missing here.

objmagic replied:
@Octachron Thanks for your review. Coming back to this comment after three months, I am still not sure what to do here because inherited field gets expanded in documentation (in elsewhere as well, e.g. toplevel display). This means when we encounter an inherited field (`Oinherit of core_type`), we want to fetch documentation associated with it somewhere, which means we need to save doc generation result for all object type declaration. I am not sure how to handle cross-module/file case either. Any suggestion is welcomed here...

Octachron replied:
@objmagic, sorry for the delayed answer. I will try to have by a look by the end of the month to see if it is possible to implement documentation comment for inherited fields in a reasonable way within ocamldoc. However, I don't think that non-optimal support within ocamldoc should block integration ((inlined record had a quite imperfect support for a time, and documentation comment can still not be attached to polymorphic variants)).

yallop commented:
> One problem I am aware of is that the following code
> 
> ```ocaml
> class type t1 = object method f : int end
> type t = < #t1 >
> ```
> will give you error message: "Error: The type #t1 is not an object type".

That's the correct behaviour, since `#t1` is not a closed object type.  The following should be accepted, though (and apparently is):

```ocaml
class type t1 = object method f : int end
type t = < t1 >
```

objmagic committed:
Correct error message, more tests.

objmagic committed:
Update ocamldoc tests

Also add test for polymorphic variants.

objmagic committed:
Update Changes entry.

objmagic commented:
@yallop 

Thanks. This is correct behavior but we do not get correct error message. 
I have updated at https://github.com/ocaml/ocaml/pull/1118/commits/cb060cd16ef1ad303a4fd52c0cc711f204cb007c and https://github.com/ocaml/ocaml/pull/1118/commits/2ce05895e9f0eb66489297ab9ee226594a7e7050.

objmagic commented:
@Octachron 

Thank you for the review. I have updated ocamldoc tests at https://github.com/ocaml/ocaml/pull/1118/commits/f7283dafb057abc07115d296e36a6b86039da328 and https://github.com/ocaml/ocaml/pull/1118/commits/edfb8afb0e448e22cddddcc4cfb5a786a5d14cc5.
> I think it would be better to not expand inherited fields in the documentation

I agree with you. Only in documentation should they not be expanded. However I am not sure how to control this or if it is possible. 

Also notice in the case of polymorphic variants, inherited variants get expanded at everywhere.

objmagic committed:
Really correct the error message.

objmagic committed:
Missing arrow.

objmagic committed:
Merge branch 'trunk' into object-inherit

objmagic commented:
@garrigue Ping again. What do you or `caml-dev` think about this PR?

garrigue commented:
Should check it in more detail, but the feature is straightforward and the behavior reasonable.
The only question left is whether we want to include something that is already possible using `class type`. I suppose a strong argument is from Javascript users, who use object types but not classes.

Ultimately, since this is a new feature, I think we need a consensus: discuss it on caml-devel or at a physical meeting.

gasche commented:
If the code is good (and I'm not competent to review it), I would be in favor of taking the feature in. It does not add complexity (especially given that polymorphic variants already support it), and I don't find the redundancy with class types an issue here -- and I have used object types myself in situations where this would have been handy.

fpottier commented:
The feature seems generally reasonable to me.

What is the desired behavior in case two methods have the same name?
Do we get a type error? a unification of the two types of the two methods? the more recent method hiding the older one?

Also, considering that repeated method declarations are currently not allowed:
```
# type t = < m: int; m: int >;;
Error: This is the second method `m' of this object type.
       Multiple occurences are not allowed.
```
it might make sense to allow this if it is allowed indirectly (via inheritance).

objmagic commented:
@fpottier Thanks for your review. 

>it might make sense to allow this 

I agree. And I'd like to hear other people's opinion of your comments as well.

yallop commented:
When both inherited types have methods with the same names I think unifying the types of the methods is the best approach.

OCaml already performs unification when two inherited polymorphic variable types have the same tag.  For example, in the following code `'a` and `'b` are unified, since the type annotation on `x` inherits the types ``[`M of 'a]`` and ``[`M of 'b]``:

```ocaml
# fun (_ : ('a * 'b)) x -> (x : [ [`M of 'a] | [`M of 'b] ]);;
- : 'a * 'a -> [ `M of 'a ] -> [ `M of 'a ] = <fun>
```

With this PR the behaviour for object types is similar, unifying the types of the methods: 

```ocaml
# fun (_ : ('a * 'b)) x -> (x : < <m: 'a> ; <m: 'b> >);;
- : 'a * 'a -> < m : 'a > -> < m : 'a > = <fun>
```

This PR also relaxes the constraint that currently prohibits having multiple occurrences of methods with the same name.  In OCaml 4.05.0:

```ocaml
# type 'a t = [`M of 'a | `M of int];;
type 'a t = [ `M of 'a ] constraint 'a = int
# type 'a t = <m: 'a ; m: int>;;
Characters 12-28:
  type 'a t = <m: 'a ; m: int>;;
              ^^^^^^^^^^^^^^^^
Error: This is the second method `m' of this object type.
       Multiple occurrences are not allowed.
```

With this PR:

```ocaml
# type 'a t = <m: 'a ; m: int>;;
type 'a t = < m : 'a > constraint 'a = int
```

objmagic committed:
merge trunk and update test

objmagic committed:
add tests

objmagic committed:
fix new test

objmagic commented:
Yes I just realized this has been implemented. I have updated tests.

damiendoligez commented:
> This PR also relaxes the constraint that currently prohibits having multiple occurrences of methods with the same name.

I'd like to hear what @garrigue has to say about this point.

garrigue commented:
If we seek symmetry with polymorphic variants, I think it is natural to remove this constraint.

garrigue reviewed (approved):
The implementation is clean, and in clear symmetry with polymorphic variants.
If nobody objects, I propose merging this before 4.06.

gasche commented:
I think that merging earlier is better than later to avoid conflicts. I'm planning to go ahead and merge in a few days if nobody does it before or objects.

I checked that the CI failures are not an issue (they are failure in Camlp4 currently not building under trunk).

objmagic commented:
@gasche Sounds good. I agree that if this PR looks all good, merge it sooner so that we don't need to deal with more merge conflicts.

gasche commented:
I made a mistake when assessing the travis CI results: it is indeed this PR that is breaking Camlp4, and this should be fixed. In the meantime, #1265 proposes to temporarily disable camlp4 building to avoid breaking other PRs as well.

gasche commented:
See ocaml/camlp4#123.
